pkg/cache/context.go

```diff
@@ -1727,13 +1729,21 @@ func (ctx *Context) finalizePods(existingPods []*v1.Pod) error {
 	// convert the pod list into a map
 	podMap := make(map[types.UID]*v1.Pod)
 	for _, pod := range pods {
+		// if the pod is terminated finalising should remove it if it was running in register
+		if utils.IsPodTerminated(pod) {
+			continue
+		}
 		podMap[pod.UID] = pod
 	}
 
-	// find any existing nodes that no longer exist
+	// find any existing pods that no longer exist
 	for _, pod := range existingPods {
+		// skip if the pod was already terminated during register
+		if pod == nil {
+			continue
+		}
 		if _, ok := podMap[pod.UID]; !ok {
-			// node no longer exists, delete it
+			// pod no longer exists, delete it
```

Comment:
Just curious. In this phase we have added event handler, so those pods may be removed already from `context` already. It seems to me that is a race condition but it is fine as `ctx.DeletePod(pod)` is no-op for repeated delete. If this description is valid, maybe we should add comments for it. Also, we should add tests for `DeletePod` to make sure it is no-op in repeated deletes.

Reply:
I'm planning to enhance test coverage for `Context`, we can have a separate test case for `DeletePod()` to make sure that's idempotent. But we can do this in this PR as well.

Reply:
+1 to address that in separate PR :)

Reply:
>  so those pods may be removed already

The _maybe_ is what we're trying to catch. It could be that the pod was removed between the list in registerPods and the event handler add later on. It will not show here but the event handler has not seen it being removed. Those we want to catch. Not the ones that have already been deleted by the handlers. We should not have to describe that in this function.

> Also, we should add tests for DeletePod to make sure it is no-op in repeated deletes.

See above for the comment made by Peter around the unit test coverage of the context code. It needs to be extended. Any code path or function that declares to be idempotent should have tests for it.